dvc/remote/local.py

```diff
@@ -202,8 +216,11 @@ def isdir(self, path_info):
     def walk(self, path_info):
         return os.walk(path_info["path"])
 
+    def get_hash_list(self):
```

Comment:
what is the point of this wrapper?

Reply:
to reuse the hash list inside output.base

Reply:
So why not just set `self.hash = [self.PARAM_CHECKSUM]` in the RemoteBase and use it instead of `get_hash_list` and `get_prefer_hash_type`?

Reply:
From my practice (in other languages), functions or methods are more flexible/optimizable/better than class fields in many cases.

1. Most important thing is to prevent `hash` get changed from outside.

2. Like `get_prefer_hash_type` will be more clear and easy to refactor than self.hash[0] in the code(11 callers) in the future, and the callers don't have to figure out the differences between different remote implements, on the other hand, other remote implements are free to use other different strategies.

3. `set self.hash = [self.PARAM_CHECKSUM] in the RemoteBase` will need long-term memory in heap for gc process to check, instead of short-term or stack-alloc(if compiler is smart enough) here for other implements.

Reply:
Great points! getters/setters are nice when you have some complex logic, but in our case properties are more than enough. 

Thinking about it a little more, it actually feels like much of that checksum logic should be contained in the `dvc/remote/local`. That way you won't even have to access those hash types from the outside. For example, Stage.changed_checksum() could then look like:
```
return self.checksum != self._compute_checksum()
```

and Stage.compute_checksum() would
```
...
return self.cache.local.dict_checksum(d, exclude=...)
```

that way almost all of the internals are hidden where they matter.

Reply:
It is actually pretty similar to how we handle `RemoteLOCAL.cache_types`.

Reply:
refactoring so much lines for showing properties are more than enough?

Reply:
@vyloy Sorry, what do you mean?

Reply:
Refactoring this two places to avoid use property `self.hash = [self.PARAM_CHECKSUM]` outside remote is fine, but how about other places, they need to be consistent, right?

for example stage.py is_cached:
```python
        for out in outs:
            out.pop(self.repo.cache.local.get_prefer_hash_type(), None)
            out.pop(RemoteS3.PARAM_CHECKSUM, None)
```